### DISCUSSION TITLE
How to design a dependency limited to a specific actor?

### DISCUSSION THREAD

**Discussion by okmaybenot** [pointfreeco/swift-composable-architecture#3209] [General]
Context:
I'm using complete concurrency checking enabled and want to specify a dependency which is confined to the MainActor

`extension MyDependency: DependencyKey {
    
    static var liveValue: MyDependency {
        
        let viewModel = DependencyClass()
        return MyDependency(
            create: viewModel.create,
            clear: viewModel.clear,
            process: viewModel.process,
            uiState: viewModel.uiState,
            navigation: viewModel.navigation
        )
    }

Problem:
I receive the following warning:
`Main actor-isolated static property 'liveValue' cannot be used to satisfy nonisolated protocol requirement`

Which is completely logical, but I don't know how to work around this warning.
Any ideas?

Thanks!

**Comment by mbrandonw** [comment]
Hi @romanatexn, typically one does not make the dependency _itself_ an actor, main or otherwise, and instead one uses an actor under the hood of the dependency to power it. If you can provide a very simple, concrete example of a `@MainActor` dependency you are trying to register with the library I can give you a more concrete example of how this looks in practice.

**Reply by okmaybenot** [reply]
Hi @mbrandonw ! Thanks for the answer, essentially I want to do something like this:

```
@MainActor
class ViewModel {
    enum Event { case onTask }
    func process(_ event: Event) {}
}


struct ViewModelDependency : Sendable {
    var process: @Sendable (_ event: ViewModel.Event) async -> ()
}

extension ViewModelDependency: DependencyKey {
    static var liveValue: ViewModelDependency {
        let vm  = ViewModel()
        return ViewModelDependency(
            process: { e in await vm.process(e) }
        )
    }
}
```
It works and checks, but the problem is I'd like to minimize possible context switching and somehow convince compiler that ViewModelDependency will be used exclusively from MainActor and provide this signature

```
struct ViewModelDependency : Sendable {
    var process: @Sendable (_ event: ViewModel.Event) -> ()
}
```

**Reply by Narsail** [reply]
Naive question: If you want it to only work with the MainActor, isn't using `Sendable` the wrong approach here? And couldn't you just declare your process function with @MainActor?
```
struct ViewModelDependency {
    var process: @MainActor (_ event: ViewModel.Event) -> ()
}
```

**Reply by okmaybenot** [reply]
Tried it, but still got warnings:
if I don't mark ViewModelDependency as Sendable I get 
`Type 'ViewModelDependency' does not conform to the 'Sendable' protocol` (so I shouldn't just return it from static var liveValue)

If I mark it as Sendable I get
`Stored property 'process' of 'Sendable'-conforming struct 'ViewModelDependency' has non-sendable type '@MainActor (ViewModel.Event) -> ()`

**Reply by pro100filipp** [reply]
> one does not make the dependency itself an actor, main or otherwise, and instead one uses an actor under the hood of the dependency to power it

Could you please tell why this is the preferred style? I'm using actors in dependencies right now but do not see (or probably am not aware of) any problems it can cause.

**Reply by mbrandonw** [reply]
Hi @pro100filipp, what specifically do you mean by "using actors in dependencies"? The actual dependency itself is an actor, or you are using an actor inside the implementation of a dependency?

**Reply by pro100filipp** [reply]
I am using it like this
```swift
import Dependencies
import Models
import SwiftData

public protocol DataController: Sendable {
    var container: ModelContainer { get }

    func createCategory(title: String) async -> PersistentIdentifier
    func deleteCategory(id: PersistentIdentifier) async
}

actor SwiftDataController: DataController {
    let container: ModelContainer
    private let context: ModelContext

    init() {
        let schema = Schema([
            Video.self, Category.self,
        ])
        let modelConfiguration = ModelConfiguration(schema: schema, isStoredInMemoryOnly: false)

        do {
            container = try ModelContainer(for: schema, configurations: [modelConfiguration])
            context = ModelContext(container)
        }
        catch {
            fatalError("Could not create ModelContainer: \(error)")
        }
    }

    func createCategory(title: String) async -> PersistentIdentifier {
        let category = Category(title: title)
        context.insert(category)
        
        try! context.save()
        
        return category.id
    }

    func deleteCategory(id categoryId: PersistentIdentifier) async {
        guard let category = context.model(for: categoryId) as? Models.Category else { return }

        context.delete(category)
        do {
            try context.save()
        }
        catch {}
    }
}

private enum DataControllerKey: DependencyKey {
    static let liveValue: any DataController = SwiftDataController()
}

extension DependencyValues {
    public var dataController: any DataController {
        get { self[DataControllerKey.self] }
        set { self[DataControllerKey.self] = newValue }
    }
}
```

As I understand, here `DataController` protocol _describes dependency_ and its conformances are _actual dependencies_. Maybe my understanding is wrong.

**Reply by mbrandonw** [reply]
This is the problematic line in your code:

```swift
private enum DataControllerKey: DependencyKey {
    static let liveValue: any DataController = SwiftDataController()
}
```

I'm not sure how this is compiling for you, but you cannot construct a `SwiftDataController` without an async context since it is an actor.

And this is why we do not think the dependency client itself should be an actor. Dependencies need to be accessible immediately and synchronously. We cannot perform async work to construct them. But, you can make a `DataController` conformance that is a simple struct with `async` methods that uses an actor on the inside. And of course you will need some kind of `initialize() async` endpoint to create the actor.

**Reply by pro100filipp** [reply]
Hmm, this sounds rather strange to me. As far as I know, actors' initializers are synchronous – at least they do not require any async context to be run ([example](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/concurrency/#Actors) from official docs). I've been using actors like this everywhere, and you can try some simple actor example and compile it yourself – I am sure it **will** compile.
Moreover, according to this [proposal](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0327-actor-initializers.md), initializers really **are** synchronous, thus leading to some restrictions on what can be done inside them. And of course, one can make an asynchronous initializer to satisfy one's needs. Maybe I do not know about some hard underneath implementations of actor initialization mechanism that implicitly leads to asynchronous initialization – if so, I'd be very thankful to know about this.

**Reply by mbrandonw** [reply]
Hi @pro100filipp, apologies! I was thinking of a _global_ actor, not a regular actor. Global actors require isolation in order to instantiate. If your code compiles just fine with full concurrency warnings / Swift 6 mode, then I say go for it!

**Reply by pro100filipp** [reply]
Ahaha, nice! Because I've already started thinking I was going crazy. By the way, I was pretty close to thinking of using a global actor for a dependency and was wondering how to do this. So, thank you for the tip about wrappers!

**Reply by pro100filipp** [reply]
@romanatexn answering your question
```swift
@MainActor
class ViewModel {
    enum Event { case onTask }
    func process(_ event: Event) {}
}

struct ViewModelDependency {
    var process: @MainActor @Sendable (_ event: ViewModel.Event) -> ()
}

extension ViewModelDependency: DependencyKey {
    static var liveValue: ViewModelDependency {
        let vm  = ViewModel()
        return ViewModelDependency(
            process: { e in vm.process(e) }
        )
    }
}
```

**Reply by tdrraluca** [reply]
@mbrandonw Can you please elaborate a bit more on the point below and that was previously mentioned in this thread?

> you will need some kind of initialize() async endpoint to create the actor

I have been looking at [The Web Sockets case study](https://github.com/pointfreeco/swift-composable-architecture/blob/bf0d2aca4b41aa90d87711edd0d394bd68c7140c/Examples/CaseStudies/SwiftUICaseStudies/03-Effects-WebSocket.swift), but the approach used in this case study consists of accessing the global actor shared instance from the async context of dependency closures. I cannot wrap my head around how adding an initialize() async endpoint can enable the dependency closures to share the same instance of a type that has an actor isolated initialiser.

Many thanks!

**Comment by kamcma** [comment]
[The Web Sockets case study](https://github.com/pointfreeco/swift-composable-architecture/blob/bf0d2aca4b41aa90d87711edd0d394bd68c7140c/Examples/CaseStudies/SwiftUICaseStudies/03-Effects-WebSocket.swift) provides an example of how a dependency can _wrap_ an actor, rather than _being_ the actor. In this case, it is its own global actor, but there are lots of things you could do here, including using the main actor.

**Reply by okmaybenot** [reply]
Yes, but it's also using context switching, meaning there's an await inside wrapper

essentially same as here https://github.com/pointfreeco/swift-composable-architecture/discussions/3209#discussioncomment-9884215

**Reply by kamcma** [reply]
I'm not certain what you want is possible. Dependency clients registered with swift-dependencies must be sendable; it's right [here](https://github.com/pointfreeco/swift-dependencies/blob/b074ba7788205b09b14a792dd059ae54a7aa540e/Sources/Dependencies/DependencyKey.swift#L123) in the `DependencyKey` protocol. If you want to interact with the main actor in the implementation of a dependency, yes, you will incur a thread-hop. It's just the API design of the swift-dependencies library. Can you explain why a thread-hop is not acceptable? I believe the thread hop is actually squashed away in TCA tests because the TestStore makes use of the `withMainSerialExecutor` tool brought in transitively from swift-concurrency-extras.

A possible alternative you could explore. Fall back to the pre-swift-dependencies style of dependencies: the environment. Make a client interface that you do _not_ register with swift-dependencies, and are thus unconstrained to mark the whole thing MainActor if you wish. You hold on to this client as plain stored property on the reducer. You know the store lives on the main actor, so the compiler should be happy to let you interact with it. You lose some of the ergonomic niceties of swift-dependencies, such as having to pass the client around explicitly, but you can still make live/test/preview implementations, just as we all did with TCA before swift-dependencies existed. I think that is the cost of making a non-sendable dependency.
